dev-docs/bookshelf-sigv4-flow.txt

```diff
@@ -0,0 +1,50 @@
+Bookshelf Sigv4 Authentication
+Basic High-Level Flow
+Lincoln Baker lbaker@chef.io
+
+PURPOSE
+
+The purpose of this document is to capture the high-level flow for Bookshelf sigv4 authentication and present a 'basic scaffolding.'  The depicted flow is purposefully somewhat simplified or idealized and is not meant to capture every single eye-glazing detail.  The assumption is that if the developer can understand the high-level flow, details can be dug into and understood at one's leisure.
+
+MODULES
+
+The two most relevant modules for understanding Bookshelf's sigv4 authentication flow are:
+bksw_wm_base.erl
+bksw_sec.erl
+For our purposes here, the flow starts in bksw_wm_base, which at some point calls bksw_sec.
+
+BKSW_WM_BASE
+
+The bulk of what bksw_wm_base does is parse or disassemble the request, and then determine whether the request is malformed, all via malformed_request/2.  'Malformed' basically means the request can't be parsed, has erroneous or nonexistent constituent parts, etc.  If the request is malformed, access is denied.  Otherwise, authentication proceeds with bksw_sec:is_authorized/2.
+
+malformed_request/2 branches by matching on two function clause heads:
+
+1) If auth_check_disabled = true, the request is deemed NOT malformed.  This is basically used to bypass checking whether the request is malformed.

```

Comment:
Could you add a note as to when we might need/want to bypass authn like this?

Reply:
I'm happy to add it, but I'm not sure I know the answer.  The original sigv2 code had this, so I simply carried it forth into sigv4.  Perhaps it is for the odd testing case one wants to run?

Reply:
Done.